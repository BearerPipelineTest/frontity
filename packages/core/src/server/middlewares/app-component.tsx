import { Middleware, Next } from "koa";
import { Context } from "@frontity/types";
import { FilledContext } from "react-helmet-async";
import App from "../../app";

/**
 * Defines the App to be used for rendering.
 *
 * @param ctx - Koa context
 */
export const appComponent = async (
  ctx: Context,
  next: Next
): Promise<Middleware> => {
  const { store } = ctx.state;

  ctx.state.helmetContext = {} as FilledContext;

  ctx.state.App = () => (
    <App store={store} helmetContext={ctx.state.helmetContext} />
  );

  // Get the correct template or html if none is found.
  if (store.libraries.frontity.App) {
    const UserDefinedApp = store.libraries.frontity.App;
    const FrontityApp = ctx.state.App;
    ctx.state.App = () => <UserDefinedApp App={FrontityApp} />;
  }

  return await next();
};
