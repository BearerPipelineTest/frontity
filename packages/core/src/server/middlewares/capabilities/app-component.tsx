import { FilledContext } from "react-helmet-async";
import App from "../../../app";

/**
 * Defines the App to be used for rendering.
 *
 * @param store - The store.
 * @param ctx - Koa context.
 */
export const appComponent = (store, ctx) => {
  ctx.state.helmetContext = {} as FilledContext;

  // Set the App.
  store.libraries.frontity.App = function FrontityApp() {
    return <App store={store} helmetContext={ctx.state.helmetContext} />;
  };
};
