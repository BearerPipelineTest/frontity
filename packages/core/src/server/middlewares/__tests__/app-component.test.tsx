import { Next } from "koa";
import { appComponent } from "../app-component";

const createStore = (extended = {}) => ({
  libraries: {
    frontity: {
      ...extended,
    },
  },
});

const createContext = (extended?: any) => ({
  state: {
    store: createStore(extended),
  },
});

const fakeNext: Next = async () => {
  await Promise.resolve();
};

describe("appComponent middleware", () => {
  it("should define a App on context", async () => {
    const ctx: any = createContext();

    // Call the middleware.
    await appComponent(ctx, fakeNext);

    expect(ctx.state.helmetContext).toBeDefined();
    expect(ctx.state.App.toString()).toMatchSnapshot();
  });

  it("should be able to overwrite the default App", async () => {
    const ctx: any = createContext({
      App: () => {
        return null;
      },
    });

    // Call the middleware.
    await appComponent(ctx, fakeNext);

    expect(ctx.state.helmetContext).toBeDefined();
    expect(ctx.state.App.toString()).toMatchSnapshot();
  });
});
