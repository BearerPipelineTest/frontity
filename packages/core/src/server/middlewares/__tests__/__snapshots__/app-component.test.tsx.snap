// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`appComponent middleware should be able to overwrite the default App 1`] = `
"function UserDefinedFrontityApp() {
            return jsx_runtime_1.jsx(UserDefinedApp, { App: FrontityApp }, void 0);
        }"
`;

exports[`appComponent middleware should define a App on context 1`] = `
"function FrontityApp() {
        return jsx_runtime_1.jsx(app_1.default, { store: store, helmetContext: ctx.state.helmetContext }, void 0);
    }"
`;
